gtk/tests_backend/probe.py

```diff
@@ -1,29 +1,63 @@
 import asyncio
+import contextlib
 
 import toga
 from toga_gtk.libs import GTK_VERSION, GLib, Gtk
 
 
 class BaseProbe:
-    def repaint_needed(self):
-        if GTK_VERSION < (4, 0, 0):
-            return Gtk.events_pending()
-        else:
-            return GLib.main_context_default().pending()
+    def _queue_draw(self, data):
+        widget, event = data
+        widget.queue_draw()
+        event.set()
 
     async def redraw(self, message=None, delay=0):
         """Request a redraw of the app, waiting until that redraw has completed."""
-        # Force a repaint
-        while self.repaint_needed():
+        if (
+            hasattr(self, "native")
+            and self.native
+            and hasattr(self.native, "queue_draw")
+        ):
+            draw_queued = asyncio.Event()
+            GLib.idle_add(self._queue_draw, (self.native, draw_queued))
+            await draw_queued.wait()
+
+            if frame_clock := self.native.get_frame_clock():
+                handler_id = None
+                with contextlib.suppress(asyncio.TimeoutError):
```

Comment:
This makes me a little nervous - what would cause a repaint to take >0.05s? and what happens if it does?